### ISSUE TITLE
`T ?? null` could be lowered to `T`

### ISSUE THREAD

**Issue by xtqqczze** [dotnet/roslyn#55919]
This could perhaps be achieved by adding code like the following to [LocalRewriter_NullCoalescingOperator.cs](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs)

```csharp
if (rewrittenLeft.Type.IsReferenceType && RemoveIdentityConversions(rewrittenRight).IsDefaultValue())
{
    return rewrittenLeft;
}
```

This optimization does not seem to be being made by the JIT:
```csharp
object M(object o) => o ?? default;
object N(object o) => o;
```

```asm
C.M(System.Object)
    L0000: test rdx, rdx
    L0003: jne short L0007
    L0005: xor edx, edx
    L0007: mov rax, rdx
    L000a: ret

C.N(System.Object)
    L0000: mov rax, rdx
    L0003: ret
```

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAZgAJiAmcgYXIG8BYAKHPcoomACsYwM5ALIAKbnwHkIASkasOCygHYp5APxryAExgAzbAFcANhgDc8jgF8L7MlN79BAOTEPJMuW0W2VEc1/ZrFksgA==

**Comment by xtqqczze** [comment]
> It feels like SDK analyzers or warning wave could also recommend the user to simplify the `<left> ?? <constant-null-expr>` expression to `<left>`.

Yes, once the code quality issue has been fixed in the compiler we could introduce a code style analyzer rule like: 

    Style -> Unnecessary code rules -> Remove unnecessary null-coalescing operator (IDEXXXX)

**Comment by xtqqczze** [comment]
Any fix should also consider the null-coalescing assignment operator:

```csharp
object Before(object o) => o ??= default;
object After(object o) => o;
```

**Comment by CyrusNajmabadi** [comment]
I don't think it makes sense for the compiler to do anything here.  The code is legit just strange.  If this code was written, it would be better for an analyzer to say "hey... don't do that, write this instead: ..."

**Comment by AlekseyTs** [comment]
I agree with @333fred here. It looks that this is an optimization for a pathological case. Base on that, I'd prefer we don't try to add it to the compiler and avoid unnecessary complexity and risk.

**Comment by 333fred** [comment]
I think you meant @CyrusNajmabadi, but I'm also ok with that conclusion.

**Comment by AlekseyTs** [comment]
> I think you meant ...

Thanks! Corrected.
